Omni token gate

cliffhall opened the pull request:
Add support for standard ERC721, ERC20, and ERC1155 tokens as conditional token for gate check, while not breaking anything with existing implementation.

- `Gate` refactored 
  -  Events declared in the contract rather than interface
  - term is **conditional token** rather than **non-transferable token** 
  - `TokenType` is an enum TOKEN or MULTI_TOKEN (ERC20/721 = TOKEN, ERC115.= MULTI_TOKEN)
  - extra constructor arg called `_conditionalTokenType` (no setter/getter yet)
  - `setConditionalContract` setter takes contract address and type
  -  `getConditionalContract` getter returns contract address and type
  - interfaces for Token and MultiToken types that have the proper `balanceOf` signature
  - check function now calls `balanceOf` appropriately based on `conditionalTokenType`
 
- `IGate` refactored
  - Events moved to contract
  - Unnecessary function signatures removed, i.e., everything not called by the other contracts

- Gate unit tests still work
- Conditional Commit unit tests work with all conditional token types 
    - ERC20
    - ERC721
    - ERC1155
    - ERC1155NonTransferrable (Quest)

cliffhall committed:
WIP: Omni-token gate

* Refactor gate contract and associated collaborators, interfaces, and tests to expect a 'conditional token' rather than the non-transferrable 'quest token'.

* The constructor of the gate takes a new tokenType argument (Token/MultiToken), which is used in the check method to determine how to call the balanceOf method.

* Also the nftTokenId is not always needed, only in the case of an 1155 (multitoken). For ERC20 and ERC721, we only need the user's wallet address and they will either have a balance on the contract or not.

* Contracts compile, but still need to get tests working with new tokenType argument

cliffhall committed:
WIP: Get unit tests running, trim interface

- All tests run, but no new tests as yet for ERC721/ERC20 as the conditional token

* In IGate.sol,
  - Refactor / move event declarations to Gate contract.
  - Remove unnecessary methods for gate clients

* In Gate.Sol,
  - Refactor / add event declarations from Gate contract.
  - Remove override from methods no longer referenced by IGate.sol

* In constants.ts,
  - added TOKEN_TYPE constant
    - TOKEN
    - MULTI_TOKEN

* In 11_gate.ts and 12_conditional_commit.ts,
  - added TOKEN_TYPE.MULTI_TOKEN as third constructor arg in all deployments of gate.

*

zajck reviewed (commented):
Looks good to me :+1: 

In BosonRouter.sol there are still a few places with reference to `ERC115NonTransferrable` instead of `conditional`
Lines 213, 298, 402, 487

cliffhall commented:
Thanks @zajck. I was focuse on the gate and didn't notice (or expect) references to the nontransferable there. Will clean up this morning.

cliffhall committed:
* In BosonRouter.sol,
  - Replaced all mentions of "NFT (ERC115NonTransferrable)" with "token of type ERC721, ERC1155, or ERC20."

NFTanaconda commented:
cool

cliffhall committed:
WIP: Accessors and unit tests

* In Gate.sol,
  - Added conditionalTokenType param to LogConditionalContractSet event
  - Refactored setConditionalTokenContract function to take _conditionalTokenType as an argument and set it.
  - Refactored getConditionalTokenContract to return conditional token address and type

* Added SampleERC20.sol, SampleERC721.sol, SampleERC1155.sol

* In 11_gate.ts,
  - Added SampleERC20, SampleERC721, SampleERC1155 to typechain imports
  - Added / initialized SampleERC20_Factory, SampleERC721_Factory, and SampleERC1155_Factory
  - Added contractSampleERC20, contractSampleERC721, and contractSampleERC1155
  - In deployContracts,
    - Moved .deployed calls to right after their associated .deploy call. It happens fast in tests, but in a live network they would possibly have issues when there are dependencies on eariler deployed contracts
    - Deployed contractSampleERC20, contractSampleERC721, and contractSampleERC1155
    - in all setConditionalTokenContract calls,
      - added constants.TOKEN_TYPE.MULTI_TOKEN (currently just doing the nontransferrable 1155 still)
      - added test '[NEGATIVE][setConditionalTokenContract] Should revert if supplied invalid token type'

* In 12_conditional_commit.ts,
  - Added SampleERC20, SampleERC721, SampleERC1155 to typechain imports
  - Added / initialized SampleERC20_Factory, SampleERC721_Factory, and SampleERC1155_Factory
  - Added contractSampleERC20, contractSampleERC721, and contractSampleERC1155
  - In deployContracts,
    - Deployed contractSampleERC20, contractSampleERC721, and contractSampleERC1155
    - in all setConditionalTokenContract calls,
      - added constants.TOKEN_TYPE.MULTI_TOKEN (currently just doing the nontransferrable 1155 still)

cliffhall committed:
WIP: Run prettier on tests

cliffhall committed:
WIP: Update conditional commit tests

* Added mock token contracts (basic implementations with mint methods)
  - SampleERC20.sol
  - SampleERC721.sol
  - SampleERC1155.sol

* In 12_conditional_commit.ts,
  - a little housecleaning on globally available variables, constants, and functions
  - refactor/extracted some constants and functions from within describe scope to global scope
  - added factories for sample token contracts
  - added sample token contract deployment to deployContracts method
  - refactor/extracted minting of non-transferrable token to globally available setupConditionalToken function, which takes a token argument indicating which conditional token should be used, configuring the gate and minting a the appropriate token to the user
  - added an array of objects describing token specific test runs, each with a displayable name and a token number
  - In the two top-level describe blocks,
    - loop over the runs array,
    - in beforeEach,
      - refactor/extracted from of some nested beforeEach definitions
        - call deployContracts
        - call preparePromiseKey
      - call setupConditionalToken (passing the current run's token number)

zajck commented on test/12_conditional_commit.ts:
```diff
@@ -38,379 +63,507 @@ let BosonRouter_Factory: ContractFactory;
+                      constants.QTY_10,
+                    ],
+                    contractGate.address,
+                    constants.NFT_TOKEN_ID,
```
It would be nice to rename `NFT_TOKEN_ID` to `CONDITIONAL_TOKEN_ID` to be in line with other changes (applies in other places as well)

hswopeams replied:
Same in the `it `

cliffhall replied:
Good call. I replaced `NFT_TOKEN_ID` with `CONDITIONAL_TOKEN_ID` throughout the codebase which included `constants.ts`, `10_erc155_non_transferrable.ts`, `11_gate.ts`, and `12_conditional_commit.ts`

zajck commented on test/12_conditional_commit.ts:
```diff
@@ -38,379 +63,507 @@ let BosonRouter_Factory: ContractFactory;
+        constants.ONE,
+        constants.ZERO_BYTES
+      );
+      //console.log("Setup Quest");
```
comment can be removed

cliffhall replied:
Had one of these for all the token types. Removed them all

zajck commented on test/12_conditional_commit.ts:
```diff
@@ -38,379 +63,507 @@ let BosonRouter_Factory: ContractFactory;
+
+      // Mint an ERC20 token for buyer
+      await contractSampleERC20.mint(users.buyer.address, constants.ONE);
+      //console.log("Setup ERC20");
```
comment can be removed

zajck commented on test/12_conditional_commit.ts:
```diff
@@ -38,379 +63,507 @@ let BosonRouter_Factory: ContractFactory;
+        constants.NFT_TOKEN_ID,
+        constants.ONE
+      );
+      //console.log("Setup ERC1155");
```
comment can be removed

zajck commented on test/12_conditional_commit.ts:
```diff
@@ -38,379 +63,507 @@ let BosonRouter_Factory: ContractFactory;
+
+      // Mint an ERC721 token for buyer
+      await contractSampleERC721.mint(users.buyer.address);
+      //console.log("Setup ERC721");
```
comment can be removed

zajck reviewed (changes requested):
Very elegant solution for all different token types.
But if I'm not mistaken, some of the tests inside a loop does not really depend on the token type, therefore the same tests are performed 4 times instead of just once. This happens for all cases when testing the creation of vouchers (`TOKEN SUPPLY CREATION WITH TOKEN CONDITIONAL COMMIT (Create Voucher Set)`).

Second loop inside `VOUCHER CREATION (Commit to buy)` seems ok.

hswopeams commented on test/12_conditional_commit.ts:
```diff
@@ -447,280 +602,295 @@ describe('Create Voucher sets and commit to vouchers with token conditional comm
+          });
+
+          describe('Flow with automatic gate.registerVoucherSetId', () => {
+            it('Should be able to create Voucher with gate address and non empty nft token id', async () => {
```
Change to conditional token Id

cliffhall replied:
Replaced "nft token id" with "conditional token id" in several locations throughout the contract.

hswopeams commented on test/12_conditional_commit.ts:
```diff
@@ -772,306 +944,301 @@ describe('Create Voucher sets and commit to vouchers with token conditional comm
+          });
+
+          describe('Flow with automatic gate.registerVoucherSetId', () => {
+            it('Should be able to create Voucher with gate address and non empty nft token id', async () => {
```
Change to conditional token Id

hswopeams commented on test/12_conditional_commit.ts:
```diff
@@ -1121,270 +1290,251 @@ describe('Create Voucher sets and commit to vouchers with token conditional comm
+          });
+
+          describe('Flow with automatic gate.registerVoucherSetId', () => {
+            it('Should be able to create Voucher with gate address and non empty nft token id', async () => {
```
Change to conditional token Id

hswopeams reviewed (changes requested):
Looks pretty good. I have a few questions/comments/requests for change.

cliffhall commented:
> But if I'm not mistaken, some of the tests inside a loop does not really depend on the token type, therefore the same tests are performed 4 times instead of just once. This happens for all cases when testing the creation of vouchers (`TOKEN SUPPLY CREATION WITH TOKEN CONDITIONAL COMMIT (Create Voucher Set)`).
> 
> Second loop inside `VOUCHER CREATION (Commit to buy)` seems ok.

@zajck  Yes, this was my expectation as well, but I wasn't 100% certain if there was anything in those voucher set creation tests that would possibly break. So erring on the side of caution, I put the loop on those tests as well. I will take them out.

cliffhall commented on contracts/Gate.sol:
```diff
@@ -2,35 +2,81 @@
+        uint256 indexed _tokenIdSupply
+    );
+
     mapping(uint256 => uint256) private voucherToToken;
```
Another good catch. Yes it was called `voucherToToken` before, but is using the voucher set's token id for the mapping to the conditional token. Changing to `voucherSetToToken`.

cliffhall commented on contracts/mocks/SampleERC1155.sol:
```diff
@@ -0,0 +1,28 @@
+ * @title Sample ERC1155 NFT for Unit Testing
+ * @author Cliff Hall
+ */
+contract SampleERC1155 is ERC1155 {
```
Yes. This was copypasta from another project, should have aligned the naming, thanks.

cliffhall commented on test/11_gate.ts:
```diff
@@ -199,8 +230,9 @@ describe('Gate contract', async () => {
-    await gate.setNonTransferableTokenContract(
-      contractERC1155NonTransferable.address
+    await gate.setConditionalTokenContract(
+      contractERC1155NonTransferable.address,
```
There is actually only one test that needs to be repeated over all the token types and that's the one that calls the `Gate.check()` method. Elsewhere it checks that you can't set a zero address or an invalid token type. 

I've set up a separate `describe` block and refactor/moved the check test into it, wrapped by a loop over the token types similar to what is implemented in the conditional commit tests.

cliffhall commented on contracts/BosonRouter.sol:
```diff
@@ -1192,18 +1192,18 @@ contract BosonRouter is
+    function finalizeConditionalOrder(uint256 _tokenIdSupply, address _gateAddress, uint256 _conditionalTokenId) internal {
         voucherSetToGateContract[_tokenIdSupply] = _gateAddress;
 
         emit LogConditionalOrderCreated(_tokenIdSupply, _gateAddress);
```
Agreed. But it won't always be present.

- For ERC20, the contract is the token, so it would never have a tokenId registered
- For ERC1155, it needs a tokenId always.
- For ERC721, it's interesting because you could target
  - the owner of a **specific** NFT (Avastar number 153) 
  - holders of **any** NFT on the conditional token contract
  
 I am only testing the use of a conditional token ID on the Quest and ERC1155.

hswopeams replied:
The code change looks fine, but it looks like the test cases that check LOG_CONDITIONAL_ORDER_CREATED in 12_conditional_commit.ts weren't changed to check for the conditional token Id.

hswopeams replied:
The code changes were made, but it looks like the LOG_CONDITIONAL_ORDER_CREATED assertions in 12_conditional_commit.ts weren't updated to check the conditional token Id

cliffhall commented on test/11_gate.ts:
```diff
@@ -72,23 +82,44 @@ describe('Gate contract', async () => {
+    contractBSNTokenDeposit: MockERC20Permit,
+    contractSampleERC20: SampleERC20,
+    contractSampleERC721: SampleERC721,
+    contractSampleERC1155: SampleERC1155;
```
I added a cycle of tests over all token types that calls the `Gate.check()` method.

cliffhall committed:
Remediate issues raised in code review

* Across codebase,
  - refactor/renamed Sample* to Mock*
  - changed nftTokenID to conditionalTokenId
  - changed NFT_TOKEN_ID to CONDITIONAL_TOKEN_ID
  - changed EMPTY_NFT_TOKEN_ID to EMPTY_CONDITIONAL_TOKEN_ID

* In Gate.sol,
  - removed unused imports
  - refactor/renamed voucherToToken to voucherSetToToken
  - replaced 'non-transferable token' with 'conditional token' in comments

* In 11_gate.ts,
  - added apparatus for running tests over all conditional token types
  - the only test that actually needed to be repeated was the one that calls the Gate.check() method. Moved it to a separate describe block and looped.
  - added _conditionalTokenId to LogConditionalOrderCreated event

zajck reviewed (changes requested):
Beside the comments in code I think it would also be good at this point to update the definition of  `Gate` in `docs/contracts/overview.md`. And maybe also all other paragraphs in this document where it describes conditional commit.

hswopeams reviewed (changes requested):
Still one outstanding change that needs to include the unit test code. See comments.

zajck committed:
removed non transferrable from BosonRouter

zajck committed:
docs update

Commit:
Added param to LogConditionalOrderCreated assertions
And added LogConditionalOrderCreated assertion to a few test cases

hswopeams reviewed (approved):
Reviewed changes @zajck made according to his requests for changes so that we can get this PR approved

mischat commented:
am going to look at this over the weekend, awesome stuff!
